Code/Framework/AzCore/AzCore/Serialization/Json/JsonImporter.cpp

```diff
@@ -144,21 +159,170 @@ namespace AZ
         return ResultCode(Tasks::Import, Outcomes::Success);
     }
 
+    JsonSerializationResult::ResultCode JsonImportResolver::StoreNestedImports(const rapidjson::Value& jsonDoc,
+        ImportPathStack& importPathStack,
+        JsonImportSettings& settings, const AZ::IO::FixedMaxPath& importPath, StackedString& element)
+    {
+        using namespace JsonSerializationResult;
+
+        for (auto& path : importPathStack)
+        {
+            if (importPath == path)
+            {
+                return settings.m_reporting(
+                    AZStd::string::format("'%s' was already imported in this chain. This indicates a cyclic dependency.", importPath.c_str()),
+                    ResultCode(Tasks::Import, Outcomes::Catastrophic), element);
+            }
+        }
+
+        importPathStack.push_back(importPath);
+        AZ::StackedString importElement(AZ::StackedString::Format::JsonPointer);
+        JsonImportSettings nestedImportSettings;
+        nestedImportSettings.m_importer = settings.m_importer;
+        nestedImportSettings.m_reporting = settings.m_reporting;
+        nestedImportSettings.m_resolveFlags = ImportTracking::All;
+        ResultCode result = StoreImports(jsonDoc, importPathStack, nestedImportSettings, importElement);
+        importPathStack.pop_back();
+
+        if (result.GetOutcome() == Outcomes::Catastrophic)
+        {
+            return result;
+        }
+
+        return ResultCode(Tasks::Import, Outcomes::Success);
+    }
+
+    JsonSerializationResult::ResultCode JsonImportResolver::StoreImports(const rapidjson::Value& jsonDoc,
+        ImportPathStack& importPathStack,
+        JsonImportSettings& settings, StackedString& element)
+    {
+        using namespace JsonSerializationResult;
+
+        if (settings.m_importer == nullptr)
+        {
+            AZStd::string reportMessage = "Json Importer is nullptr for ResolveImports.";
+            if (!importPathStack.empty())
+            {
+                reportMessage += AZStd::string::format(" Skipping storing imports for file '%s'.", importPathStack.back().c_str());
+            }
+            return settings.m_reporting(reportMessage,
+                ResultCode(Tasks::Import, Outcomes::Skipped), element);
+        }
+
+        if (jsonDoc.IsObject())
+        {
+            for (auto& field : jsonDoc.GetObject())
+            {
+                if (strncmp(field.name.GetString(), JsonSerialization::ImportDirectiveIdentifier, field.name.GetStringLength()) == 0)
+                {
```

Comment:
`ResolveImports` and `StoreImports` are similar enough it may be useful to move some of this inner code to a common function if it will help prevent fixing something in one function and forgetting to make the fix/change in the other.

Reply:
Well the way the logic is interleaved makes it hard to combine them, especially with the calls to ResolvedNestedImports vs StoreNestedImports